### PR TITLE
iSCSI HA storage cluster/SAN: drbd, pacemaker, corosync, iSCSI LIO

### PR DESCRIPTION
**Pull request by ryantm**
###### Motivation for this change

drbd used to contain the kernel module code and the userspace
tools. The kernel module continues to be available in the drbd
project, but is also included with the linux kernel directly, so is
probably unnecessary to include. The userpsace tools were split into
drbd-utils, which are packaged here.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by ryantm** [comment]
Result of `nixpkgs-review pr 119904` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>7 packages failed to build:</summary>
  <ul>
    <li>libvmi</li>
    <li>qemu_xen (qemu_xen_4_10)</li>
    <li>qemu_xen-light (qemu_xen_4_10-light)</li>
    <li>qubes-core-vchan-xen</li>
    <li>xen</li>
    <li>xen-light</li>
    <li>xen-slim</li>
  </ul>
</details>
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>drbd (drbd-utils)</li>
    <li>resource-agents</li>
  </ul>
</details>

**Comment by aanderse** [comment]
ping @jslight90

File: pkgs/os-specific/linux/drbd-utils/default.nix [hunk 1]
```diff
@@ -0,0 +1,122 @@
+{ lib, stdenv
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/drbd-utils/default.nix:1]:
```suggestion
{ lib
, stdenv
```

File: pkgs/os-specific/linux/drbd-utils/default.nix [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+stdenv.mkDerivation rec {
+  pname = "drbd-utils";
+  version = "9.16.0";
+
+  src = fetchurl {
+    url = "https://linbit.com/downloads/drbd/utils/${pname}-${version}.tar.gz";
+    sha256 = "sha256:0s1dgg4nvvvxzh9qrah3n15b92h38b8m9pa43s14fdjdg11m6wk2";
+  };
+
+  # patches = [ ./pass-force.patch ./fix-glibc-compilation.patch ];
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/drbd-utils/default.nix:23]: Whats with this?

File: pkgs/os-specific/linux/drbd-utils/default.nix [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+  configureFlags = [
+    "--libdir=${placeholder "out"}/lib"
+    "--sbindir=${placeholder "out"}/bin"
+    "--localstatedir=/var"
+    "--sysconfdir=/etc"
+    "--without-distro"
+  ];
+
+  makeFlags = [
+    "SHELL=${stdenv.shell}"
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/drbd-utils/default.nix:46]: Should we use runtimeShell here?

File: pkgs/os-specific/linux/resource-agents/default.nix [hunk 4]
```diff
@@ -0,0 +1,37 @@
+{ lib, stdenv
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/resource-agents/default.nix:1]:
```suggestion
{ lib
, stdenv
```

**Comment by jslight90** [comment]
> ping @jslight90

I'm working on setting up a SAN system with these changes.  I'll update with results within the next couple weeks.

File: pkgs/os-specific/linux/drbd-utils/default.nix (hunk 3 above)
Comment (ryantm) [reply pkgs/os-specific/linux/drbd-utils/default.nix]: Looks like it wasn't necessary, so I removed it.

File: pkgs/os-specific/linux/drbd-utils/default.nix (hunk 2 above)
Comment (ryantm) [reply pkgs/os-specific/linux/drbd-utils/default.nix]: Those are old patches from the drbd derivation. Removed it.

File: nixos/modules/services/network-filesystems/drbd.nix [hunk 5]
```diff
@@ -47,18 +47,33 @@ let cfg = config.services.drbd; in
...
       after = [ "systemd-udev.settle.service" "network.target" ];
       wants = [ "systemd-udev.settle.service" ];
       wantedBy = [ "multi-user.target" ];
+      serviceConfig = {
+        Type = "oneshot";
+        RemainAfterExit = "yes";
+      };
+      preStart = ''
+        # Check the configuration file for syntax errors
+        ${pkgs.drbd-utils}/bin/drbdadm sh-nop
```
Comment (aanderse) [inline nixos/modules/services/network-filesystems/drbd.nix:64]: Would `drbdadm` allow us to run this inside the sandbox and present build time errors?

**Comment by mweinelt** [comment]
Are the drbd-utils 9.x compatible with the 8.4.x kmods that are intree? AFAIK recent 9.x versions have not been upstreamed into the Linux kernel.

**Comment by ryantm** [comment]
> Are the drbd-utils 9.x compatible with the 8.4.x kmods that are intree? AFAIK recent 9.x versions have not been upstreamed into the Linux kernel.

@mweinelt The drbd-utils 9.x [compilation options](https://github.com/LINBIT/drbd-utils/blob/2b895682d20cb3f804f4ec4b30297f67cec3827a/configure.ac#L62) to disable support for 8.3 and 8.4 so I assume that means it is backward compatible. I am in the process of testing it still though.

**Comment by ryantm** [comment]
It seems to be backward compatible in that it calls out to the correct version binary:
```
# drbdadm -v adjust all
drbdsetup-84 new-resource r0
drbdsetup-84 new-minor r0 0 0
drbdmeta 0 v08 /dev/md127 internal apply-al
drbdsetup-84 attach 0 /dev/md127 /dev/md127 internal
drbdsetup-84 connect r0 ipv4:192.168.1.2:7789 ipv4:192.168.1.2:7789 --protocol=C
```

File: pkgs/os-specific/linux/drbd-utils/default.nix [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+  ];
+
+  postPatch = ''
+    patchShebangs .
+    substituteInPlace user/v84/drbdadm_usage_cnt.c \
+      --replace '			"/lib/drbd");' \
+                '			"${placeholder "out"}/lib/drbd");'
+    substituteInPlace user/v9/drbdsetup_linux.c \
+      --replace '		ret = system("/sbin/modprobe drbd");' \
+                '		ret = system("${kmod}/bin/modprobe drbd");'
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/drbd-utils/default.nix:71]: Please dont mix tabs and spaces.

File: pkgs/os-specific/linux/drbd-utils/default.nix [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+                "${docbook_xml_dtd_45}/xml/dtd/docbook/docbookx.dtd"
+    substituteInPlace documentation/v9/drbdsetup.xml.in \
+      --replace "http://www.oasis-open.org/docbook/xml/4.5/docbookx.dtd" \
+                "${docbook_xml_dtd_45}/xml/dtd/docbook/docbookx.dtd"
+    substituteInPlace documentation/v84/drbdsetup.xml \
+      --replace "http://www.oasis-open.org/docbook/xml/4.5/docbookx.dtd" \
+                "${docbook_xml_dtd_45}/xml/dtd/docbook/docbookx.dtd"
+    substituteInPlace documentation/v84/drbd.conf.xml \
+      --replace "http://www.oasis-open.org/docbook/xml/4.5/docbookx.dtd" \
+                "${docbook_xml_dtd_45}/xml/dtd/docbook/docbookx.dtd"
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/drbd-utils/default.nix:89]: This could be simplified with for f in; do .. done.

**Comment by jslight90** [comment]
I was able to setup a highly-available SAN using these drbd changes, pacemaker, corosync, and iSCSI.  I plan to submit a pull request with the pacemaker pkg, and a module for a HA SAN (depends on these changes).

**Comment by jslight90** [comment]
One problem I ran into, the OCF files to manage resources were not in the standard OCF directory (`/usr/lib/ocf`).  I'm not sure whether copying those files into place should be a part of this PR or the HA SAN module.

File: nixos/modules/services/network-filesystems/drbd.nix (hunk 5 above)
Comment (ryantm) [reply nixos/modules/services/network-filesystems/drbd.nix:52]: Maybe! But I'd prefer to not delay this to figure that out.

File: pkgs/misc/logging/pacemaker/default.nix [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+, libuuid
+, libxml2
+, libxslt
+, pam
+, pkg-config
+, python3
+
+# Pacemaker is compiled twice, once with forOCR = true to extract its
+# OCF definitions for use in the ocf-resource-agents derivation, then
+# again with forOCR = false, where the ocf-resource-agents is provided
```
Comment (jslight90) [inline pkgs/misc/logging/pacemaker/default.nix:22]: Typos:
```suggestion
# Pacemaker is compiled twice, once with forOCF = true to extract its
# OCF definitions for use in the ocf-resource-agents derivation, then
# again with forOCF = false, where the ocf-resource-agents is provided
```

File: pkgs/os-specific/linux/drbd-utils/default.nix [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+, perl
+, docbook_xsl
+, libxslt
+, docbook_xml_dtd_44
+, docbook_xml_dtd_45
+, kmod
+
+# drbd-utils are compiled twice, once with forOCR = true to extract
+# its OCF definitions for use in the ocf-resource-agents derivation,
+# then again with forOCR = false, where the ocf-resource-agents is
```
Comment (jslight90) [inline pkgs/os-specific/linux/drbd-utils/default.nix:15]: Typos:
```suggestion
# drbd-utils are compiled twice, once with forOCF = true to extract
# its OCF definitions for use in the ocf-resource-agents derivation,
# then again with forOCF = false, where the ocf-resource-agents is
```

File: pkgs/misc/logging/pacemaker/default.nix (hunk 8 above)
Comment (ryantm) [reply pkgs/misc/logging/pacemaker/default.nix]: thanks; fixed.

**Comment by ryantm** [comment]
> One problem I ran into, the OCF files

I've resolved the issue with OCF files by building drbd and pacemaker twice. Once to extract the OCF files and once with all the combined OCF files from a derivation called `ocf-resource-agents`. It is no longer necessary to put files at `/usr/lib/ocf`.

**Comment by ryantm** [comment]
I changed the title because we are going to use this PR to update all the components and modules we are using for our HA storage cluster/SAN.

**Comment by astro** [comment]
I was able to build a working Pacemaker cluster with the packages in this closed PR.

@ryantm @jslight90 What are your future plans? Is this up for grabs?

**Comment by ryantm** [comment]
@astro Feel free to pick it up and use what we have here! Currently, I'm evaluating using Proxmox instead, since I was able to get going in almost no time.

**Comment by astro** [comment]
Interesting. My motivation is replacing a Frankenproxmox installation with something well-defined.

**Comment by ryantm** [comment]
@astro Interesting, we should talk more! I messaged you on Matrix.

**Pull request closed without merging**